Initial version of the Vert.x AMQP client

cescoffier opened the pull request:
This PR contains the initial version of the Vert.x AMQP client.

cescoffier committed:
Initial prototype for the API

cescoffier committed:
Use converters

cescoffier committed:
Workaround docgen NPE.

cescoffier committed:
Configure the git ignore list

cescoffier committed:
Configure the indentation

cescoffier committed:
Initial working version

cescoffier committed:
Improve acknowledgment support

cescoffier committed:
Allow configuring the container id in the client options.

cescoffier committed:
Improvement around the AMQP receiver options.

cescoffier committed:
add build instruction to readme

cescoffier committed:
Improve type support

cescoffier committed:
Ensure that the connection is not null

cescoffier committed:
Change logger in AMQP Usage class

cescoffier committed:
Allow sending message without body

Unclear in the AMQP spec, but it seems to be an allowed way to pass `null`

cescoffier committed:
Ensure that the AMQP Usage waits until the ProtonSender or ProtonReceiver are opened before continuing the execution.

cescoffier committed:
Remove support of Decimal types
Remove support of Symbol
Improve List support and clarify how they are written and read
add Map support
Improve null support

cescoffier committed:
setup CI

cescoffier committed:
Add build badge to readme

cescoffier committed:
add a trace to debug CI build failure

cescoffier committed:
Ensure the proton API is accessed on the right context

cescoffier committed:
Improve the reliability of the AMQP Usage class
Explicit usage of a Vert.x instance in the SenderTest.

cescoffier committed:
send the second message once the first one is acked

cescoffier committed:
send the second message once the first one is acked

cescoffier committed:
Async close methods.

cescoffier committed:
close connection on context

cescoffier committed:
Add connection metadata

cescoffier committed:
Initial implementation of ReadStream.

cescoffier committed:
Add support for virtual host and container id

cescoffier committed:
Add support for application properties
Add the ability to retrieve a receiver without handler

cescoffier committed:
Add address() method to sender

cescoffier committed:
Add support for replies

cescoffier committed:
Initial implementation of the write stream and flow control on the sender side

cescoffier committed:
Check the behavior of the Sender close and end methods

cescoffier committed:
Test that messages are received:
* after pause / resume
* when a handler is attached later

cescoffier committed:
Test closing a consumer

cescoffier committed:
Test sender flow control

cescoffier committed:
Avoid closing twice senders and receivers
Restructure test suite to avoid killing the broker.

cescoffier committed:
Switch to the event loop when closing.

cescoffier committed:
Add tests around the closing of the AMQP receiver.

cescoffier committed:
Replace reference to the bridge.

cescoffier committed:
Add an option to configure the initial credit on the consumer side

cescoffier committed:
Improve closing the clients and add support for and endHandler

cescoffier committed:
Handle runtime exception thrown by the close method

cescoffier committed:
Check that we can close the AMQP client even without connections

cescoffier committed:
Add SSL Tests

cescoffier committed:
Fix deadlock

cescoffier committed:
Renamed body methods.

cescoffier committed:
Add javadoc to the AMQP client class

cescoffier committed:
Write javadoc for AMQP receiver and AMQP sender

cescoffier committed:
Remove a useless variant of the sender method

cescoffier committed:
Rework options.

cescoffier committed:
Make AMQP Message builder polyglot

cescoffier committed:
Add javadoc for the AMQP client options

cescoffier committed:
Cleanup AMQP Message and builder

cescoffier committed:
Add license header

cescoffier committed:
Add documentation

cescoffier committed:
Update parent version, cleanup the pom.xml file

cescoffier committed:
Add javadoc for the body methods as they are specific.

cescoffier committed:
Avoid duplication

cescoffier committed:
At that point, it's not a prototype anymore

gemmellr commented on pom.xml:
```diff
@@ -0,0 +1,188 @@
+    <dependency>
+      <groupId>com.fasterxml.jackson.dataformat</groupId>
+      <artifactId>jackson-dataformat-yaml</artifactId>
+      <version>2.9.7</version>
```
is there a reason to override this from the dependencies pom setting of 2.9.8?

cescoffier replied:
Good catch.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/ReplyManager.java:
```diff
@@ -0,0 +1,215 @@
+
+  private void handleIncomingMessageReply(AmqpMessage response) {
+    // Check if we have the reply-to-message app properties
+    String correlationId = response.applicationProperties().getString(REPLY_TO_MESSAGE_PROPERTY, response.correlationId());
```
An ApplicationProperties section might not be present in the message, so this could NPE.

Is there a reason you moved it to this 'proprietary' reply-to-reply behaviour using magic application property, vs the basic message-id + correlation-id model the old bridge used?

If desiring to only use the correlation-id field, then would a less 'proprietary' alternative solution just being to reuse the correlationId from the first 'reply' message on the second 'reply to reply' message? Right now it always generates a new correlationId when sending expecting a reply, even if its a reply-to-reply, but as the original responder is actually waiting for the correlationId value it sent (whether in magic property, or correlation-id field) then why not put it in the response correlation-id again? Is the concern having multiple exchanges using one correlation-id if there are repeating replies-to-replies?

cescoffier replied:
So, I've removed the reply feature, code is much simpler now...

gemmellr commented on pom.xml:
```diff
@@ -0,0 +1,188 @@
+
+    <dependency>
+      <groupId>io.vertx</groupId>
+      <artifactId>vertx-proton</artifactId>
```
Perhaps move up alongside the other non-test dependencies?

cescoffier replied:
Done.

gemmellr commented on src/main/asciidoc/index.adoc:
```diff
@@ -0,0 +1,181 @@
+{@link examples.AmqpClientExamples#sender(io.vertx.ext.amqp.AmqpConnection)}
+----
+
+When you create a sender, you can pass the default targeted address. However, note that passing `null` is allowed as well
```
This doesnt really convey things nicely for to me.

When you create a sender link you specify the address of the target node (e.g a queue, with the address being its name). The message 'to' field address is somewhat independent and may not be set at all, could be set to the same address for info, or could even be set different than the address of the node the link it attached to. The AMQP spec doesnt define behaviour around that latter bit as its node behaviour, which is outwith the spec, but if you e.g set it to a different address than the named queue you attached to, in most cases nothing different is likely to happen based on that being there, its not an override in that regard.

An extension was defined to have null target address represent a special 'anonymous relay' node, such that one sender link to that node can carry messages for many different nodes, with their addresses conveyed in the message to field, e.g allowing sending to various queues with one sender. Thats currently the main use of the message 'to' field I am aware of in terms of specifying a destination. Otherwise, JMS senders use it to carry JMSDestination info even in the case they arent 'anonymous' senders.

See also:
https://github.com/vert-x3/vertx-proton/blob/3.6.3/src/main/java/io/vertx/proton/ProtonConnection.java#L88
https://github.com/vert-x3/vertx-proton/blob/3.6.3/src/main/java/io/vertx/proton/ProtonConnection.java#L148

gemmellr commented on src/main/asciidoc/index.adoc:
```diff
@@ -0,0 +1,181 @@
+* {@link io.vertx.ext.amqp.AmqpSender#send(io.vertx.ext.amqp.AmqpMessage)} - send the message
+* {@link io.vertx.ext.amqp.AmqpSender#send(io.vertx.ext.amqp.AmqpMessage, io.vertx.core.Handler))}
+- send the message and expect a reply
+* {@link io.vertx.ext.amqp.AmqpSender#send(java.lang.String, io.vertx.ext.amqp.AmqpMessage)} - send the message to a different address
```
Related to earlier comment above around sender creation, I think this needs to be clearer, perhaps noting specifically that it just populates or overrides the message 'to' address, mainly useful if using an 'anonymous sender'.

gemmellr commented on src/main/asciidoc/index.adoc:
```diff
@@ -0,0 +1,181 @@
+Once you have the sender and created the message, you can send it using:
+
+* {@link io.vertx.ext.amqp.AmqpSender#send(io.vertx.ext.amqp.AmqpMessage)} - send the message
+* {@link io.vertx.ext.amqp.AmqpSender#send(io.vertx.ext.amqp.AmqpMessage, io.vertx.core.Handler))}
```
You are unlikely to ever get a reply if the original message isnt actually accepted (i.e is released/rejected/modified), which this method doesn't allow knowing whether it was, so it should perhaps define what happens there.

More of a note than actual feedback. I actually dislike the built in 'reply' handling and it being enabled by default with all the extra hoops that involves jumping underneath the sender around receivers and correlation. I only added that to the old bridge since the existing API forced it. Otherwise I'd be more inclined to have that as optional functionality created to the side, or not bother since handling replies isnt actually difficult for applications IMO and gives more options/flexibility such as fixed queues/addresses, multiple responses, etc etc.

cescoffier replied:
I've discussed it with @vietj and we are going to remove the reply management for now. From my side it brings a lot of complexity and actually limits to a "1 request - 1 response" behavior.

gemmellr commented on src/main/asciidoc/index.adoc:
```diff
@@ -0,0 +1,181 @@
+* {@link io.vertx.ext.amqp.AmqpSender#send(io.vertx.ext.amqp.AmqpMessage, io.vertx.core.Handler))}
+- send the message and expect a reply
+* {@link io.vertx.ext.amqp.AmqpSender#send(java.lang.String, io.vertx.ext.amqp.AmqpMessage)} - send the message to a different address
+* {@link io.vertx.ext.amqp.AmqpSender#sendWithAck(io.vertx.ext.amqp.AmqpMessage, io.vertx.core.Handler)} - send the message and monitor its acknowledgment
```
It feels a bit lopsided that there are 3 methods that don't allow handling acknowledgement and only one that does. For me this is actually the most useful send method, even though (or because) it doesn't allow message address setting or reply handling.

cescoffier replied:
Reply removed, so things got cleaner

gemmellr commented on src/main/java/io/vertx/ext/amqp/AmqpConnection.java:
```diff
@@ -0,0 +1,108 @@
+  /**
+   * Closes the AMQP connection, i.e. allows the Close frame to be emitted.
+   * <p>
+   * For locally initiated connections, the {@link #closeHandler(Handler)} may be used to handle the peer sending their
```
The 'locally initiated' bit doesnt really make sense for client-only usage, that would have been allowing for server usage originally.

cescoffier replied:
Fixed.

gemmellr commented on src/main/java/io/vertx/ext/amqp/AmqpMessage.java:
```diff
@@ -0,0 +1,237 @@
+
+  int priority();
+
+  String id();
```
Is the intent here to encourage using strings, and only fully facilitate others later if needed? (There are 4 message-id and correlation-id types: string, uuid, binary, ulong)

cescoffier replied:
I use string because of the polyglot generation constraints. UUID is not supported in polyglot APIs.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpClientImpl.java:
```diff
@@ -0,0 +1,90 @@
+    List<Future> actions = new ArrayList<>();
+    for (AmqpConnection connection : connections) {
+      Future<Void> future = Future.future();
+      connection.close(future);
```
Dont think this overall method is safe from outwith the context thread, as these connection closes look to touch various vertx-proton bits directly.

cescoffier replied:
Yes, in `connection.close` we must switch to the context.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpConnectionImpl.java:
```diff
@@ -0,0 +1,298 @@
+                    }
+                  });
+                } else {
+                  runWithTrampoline(x -> connectionHandler.handle(conn.mapEmpty()));
```
If connect is always run on the context is the outer call doing anything?

cescoffier replied:
Fixed.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpConnectionImpl.java:
```diff
@@ -0,0 +1,298 @@
+
+            this.connection.get().open();
+          } else {
+            runWithTrampoline(x -> connectionHandler.handle(ar.mapEmpty()));
```
As above

cescoffier replied:
Fixed.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpConnectionImpl.java:
```diff
@@ -0,0 +1,298 @@
+  }
+
+  @Override
+  public AmqpConnection close(Handler<AsyncResult<Void>> done) {
```
Per earlier comment, this seems unsafe for off-context usage as it touches lots of proton objects.

cescoffier replied:
Fixed.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpConnectionImpl.java:
```diff
@@ -0,0 +1,298 @@
+        try {
+          actualConnection
+            .closeHandler(cleanup ->
+              runWithTrampoline(x -> {
```
Since its in a callback is the outer call adding anything?

cescoffier replied:
Now that close is called on context, yes the runWithTrampoline is not required anymore.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpConnectionImpl.java:
```diff
@@ -0,0 +1,298 @@
+  @Override
+  public AmqpConnection receiver(String address, Handler<AsyncResult<AmqpReceiver>> completionHandler) {
+    ProtonLinkOptions opts = new ProtonLinkOptions();
+    ProtonReceiver receiver = connection.get().createReceiver(address, opts);
```
This wouldn't be safe unless run on the context thread, which the later bit seems like you aren't requiring (the old bridge enforced it was required here).

Same note for the subsequent creation methods below.

cescoffier replied:
Fixed

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpReceiverImpl.java:
```diff
@@ -0,0 +1,309 @@
+    connection.unregister(this);
+    if (this.receiver.isOpen()) {
+      try {
+        receiver.close();
```
As earlier note around threading.

Also this doesn't even seem to look at or wait for the actual receiver closure on the wire, is that intended?

cescoffier replied:
Nope :-)

It's fixed.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpSenderImpl.java:
```diff
@@ -0,0 +1,275 @@
+      }
+    });
+
+    this.connection.runWithTrampoline(x -> sender.open());
```
Earlier bits arent safe if this isnt on the context thread, as this seems to allow for.

cescoffier replied:
Fixed.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpSenderImpl.java:
```diff
@@ -0,0 +1,275 @@
+    connection.unregister(this);
+    if (sender.isOpen()) {
+      try {
+        sender.close();
```
As earlier notes around threading.

Also this doesn't even seem to look at or wait for the actual sender closure on the wire, is that intended?

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/ReplyManager.java:
```diff
@@ -0,0 +1,215 @@
+      builder.address(replyAddress);
+    }
+
+    String origMessageId = origin.correlationId();
```
Variable should be renamed origCorrelationId, or such like.

gemmellr commented on src/test/java/io/vertx/ext/amqp/AmqpUsage.java:
```diff
@@ -0,0 +1,266 @@
+   */
+  public void produce(String topic, int messageCount, Runnable completionCallback, Supplier<Object> messageSupplier) {
+    CountDownLatch ready = new CountDownLatch(1);
+    ProtonSender sender = connection.createSender(topic);
```
Unsafe if called off the context thread.

gemmellr commented on src/test/java/io/vertx/ext/amqp/AmqpUsage.java:
```diff
@@ -0,0 +1,266 @@
+  public void consume(String topic, BooleanSupplier continuation, Runnable completion,
+                      Consumer<AmqpMessage> consumerFunction) {
+    CountDownLatch latch = new CountDownLatch(1);
+    ProtonReceiver receiver = connection.createReceiver(topic);
```
Unsafe if called off the context thread.

gemmellr commented on src/test/java/io/vertx/ext/amqp/AmqpUsage.java:
```diff
@@ -0,0 +1,266 @@
+    CountDownLatch entities = new CountDownLatch(senders.size() + receivers.size());
+    senders.forEach(sender -> {
+      if (sender.isOpen()) {
+        sender.closeHandler(x -> entities.countDown()).close();
```
Unsafe if called off the context thread.

gemmellr reviewed (commented):
Did some review and left initial comments and questions. I haven't yet looked at most of the tests.

cescoffier committed:
Version inherited for jackson-dataformat-yaml

cescoffier committed:
Avoid NPE when handling reply

cescoffier committed:
Change the ad-hoc property for the correlation id to match the one use in the Vert.x AQMP bridge

cescoffier committed:
Reorganize dependencies

cescoffier committed:
Enforce context on the connection close

cescoffier committed:
Called sender and receiver "close" methods on the right context

cescoffier committed:
Remove call to the run on context as we are sure it will be called on the right context

cescoffier committed:
Enforce the execution context

cescoffier committed:
Enforce that proton method are executed on the context

cescoffier committed:
Ensure that the sender has been closed before calling the handler

cescoffier committed:
Rename variable

cescoffier committed:
Fix license header - avoid dangling comment

cescoffier committed:
Remove reply functionality as discussed in the review

cescoffier committed:
Fix race conditions in tests

cescoffier commented:
@gemmellr Thanks for the detailed review!

I've resolved all the issues you reported. The big change it the removal of the reply feature, which simplifies a lot of things (API, behavior...)

vietj commented:
I would like to discuss an simplifying API changes.

Currently I see that you can send messages by using the chain:

1. create a client
2. open a connection
3. create a sender
4. send the message

I would like to simplify to basically do avoid having the user create a connection and let the client create the connection under the hood. In addition a client only knows a single host so there is no ambiguity about which host to connect to.

So the change would be that a client allows to create a sender and a receiver without a connection. The connection is lazily created by the client the first time it's needed and closed when it's not needed anymore, so the user flow would be:

1. create a client
3. create a sender
4. send the message

We can also have one-shot methods to directly send individual messages when no write-stream is needed or the cost of creating a sender is not important: have on the client the same send methods that are on the sender, where the send method creates a sender, uses it and then dispose it, so sending a simple message is just a single operation on the client.

I believe such features don't require much changes.

vietj commented on src/main/java/io/vertx/ext/amqp/AmqpConnection.java:
```diff
@@ -0,0 +1,108 @@
+   * @param endHandler the end handler.
+   */
+  @Fluent
+  AmqpConnection endHandler(Handler<Void> endHandler);
```
use `closeHandler` instead, that's how it is usually called when you want to be aware of the close notification

gemmellr replied:
There is a closeHandler also. I thought both were usually defined, with endHandler called for 'unexpected' closure?

vietj replied:
`endHandler` is usually used when there is a `ReadStream`, what's the purpose of this ?

gemmellr replied:
As implemented, it fires if the connection remotely closes or disconnects without you asking for it to.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/ReplyManager.java:
```diff
@@ -32,10 +32,9 @@
 
   private static final Logger LOGGER = LoggerFactory.getLogger(ReplyManager.class.getName());
-  public static final String REPLY_TO_MESSAGE_PROPERTY = "reply-to-message";
+  public static final String REPLY_TO_MESSAGE_PROPERTY = "correlation_id";
```
Not actually an issue any longer given that the implicit reply functionality has been removed (yay), but for clarity and later reference this isn't what I was talking about with my original review comment.

This simply altered the name of the 'proprietary' application property being used. It still wasn't doing what the bridge was doing, which was to use the built in message-id and correlation-id fields of the AMQP message Properties section and not defining a custom entry in ApplicationProperties which is distinct.

gemmellr commented:
@cescoffier Thanks for the updates. I'm glad to see the implicit reply stuff disappeared also.

The send method docs/javadoc didn't seem to really change, I still think they need attention per the original comments around the 'different address' stuff. I think its not clear and is very likely to be misunderstood since most of the time it wont actually do remotely what is implied by the current text (unless you happen to be using an 'anonymous' sender).

gemmellr commented:
I'm not fond of simplifications like hiding the connection, but especially dislike the idea of of 'one shot send' methods. People inevitably use these simple things without proper consideration when presented with them. We constantly end up needing to discuss the resulting performance issues, and inefficient  load it can place on clients and servers, etc etc. It also tends to complicate the impls when we then try doing things to reduce the impact of the inefficient operations when people do then [mis]use them and get into poor situations. If the alternative were massively difficult I might see more need but I don't think that's really the case.

cescoffier commented:
> The send method docs/javadoc didn't seem to really change

Let me check that, I may have miss it.

About the `address`. I'm confused about the address and the `to` field. 
So, I'm wondering if we should not remove the method taking the `address` as a parameter.

gemmellr commented:
Removing those methods would suit me.

cescoffier committed:
Cleanup the javadoc

cescoffier committed:
Remove methods from sender accepting an address.

cescoffier committed:
cleanup import statements

cescoffier committed:
Rename receiver and sender method to createReceiver and createSender
Add support for anonymous sender
Add support for dynamic receiver
Add documentation and test for request reply

cescoffier committed:
Add support for all supported headers
http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#type-header

cescoffier committed:
Add missing license header

cescoffier committed:
Add support for durable receiver

vietj commented:
The simplification is because it is asynchronous and therefore sending a message requires 2 asynchronous method calls, we have applied this in various Vert.x clients and the feedback is rather positive than negative.

The goal is not to hide the connection, the goal is to provide a shortcut for people that want to go straight to the point and just send messages.

_People inevitably use these simple things without proper consideration when presented with them_ can you elaborate why it would be more costly ?

cescoffier committed:
Introduce AMQP Sender options

cescoffier committed:
Fix AMQP Sender auto-settled
Add test

cescoffier committed:
Allow creating receivers directly on the client

cescoffier committed:
Add methods to create senders directly from the AMQP client

cescoffier commented:
@vietj I've added methods to create senders and receivers on the client itself (so no need to manage the connection explicitly).

@gemmellr I've added anonymous senders, dynamic receivers, terminus configuration on the receiver side (including a durable property), offered and desired capabilities on the receiver side. 

Still 2 questions:

* does terminus configuration make sense on the sender side? Should we configure the source or target in this case (I would say target)
* when I close a durable receiver, I'm calling detach instead of close, is it the right behavior?

gemmellr commented:
> can you elaborate why it would be more costly ?

Those comments were mainly aimed at the 'instant send' method suggestion. Creating senders isnt free, and every time this approach is offered people abuse or misuse it unwittingly, then we wend up wasting time trying making other bits smarter to try and fix the problem we only created by trying to fix one that didn't really exist.

gemmellr commented:
@cescoffier Yes terminus config makes sense on a sender. It will typically be the target that needs configuring, but it can be either for both sender and receiver (a link is always between a source and a target, its just a question which side is which based on if sending or receiving). Detaching the durable receiver link makes sense yes, if you want to come back later and resume consuming from the same 'subscription' you then can. If you don't you close it.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpConnectionImpl.java:
```diff
@@ -268,23 +271,62 @@ public AmqpConnection createReceiver(String address, AmqpReceiverOptions receive
+        List<String> desired = receiverOptions.getDesiredCapabilities();
+        List<String> provided = receiverOptions.getCapabilities();
+
+        receiver.setDesiredCapabilities(desired.stream().map(Symbol::valueOf).toArray(Symbol[]::new));
```
This is often not going to be what is desired. Most often the capabilities used are those applied to the remote terminus details (the source/target for receiver/sender respectively), and not the link (sender/receiver) itself.

I can think of a couple cases where link capabilities/properties are used in process of establishing support for certain functionality if it wasn't already established by connection-level capabilities/properties, but those aren't something applications typically do.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpConnectionImpl.java:
```diff
@@ -268,23 +271,62 @@ public AmqpConnection createReceiver(String address, AmqpReceiverOptions receive
+      source.setExpiryPolicy(TerminusExpiryPolicy.NEVER);
+      source.setDurable(TerminusDurability.UNSETTLED_STATE);
+    } else {
+      // Check if we have individual values, not the in this case the receiver won't be considered as durable and the
```
I wouldn't have this disparity in its current state. It seems unwise to allow controlling the terminus durability manually without also offering some way for requesting a detach with closed=false, since the main thing folks would want to use this for is the same, and so setting closed=true will then typically fail to achieve what they wanted. I'd suggest either get rid of the 'durable' shortcut, or make this detect the same case and act the same, or add an option to control what happens when close() is called (or add a detach method).

gemmellr replied:
..or get rid of ability to manually set the details, and just leave the durable option shortcut for now.

gemmellr commented on src/main/java/io/vertx/ext/amqp/AmqpConnection.java:
```diff
@@ -79,19 +79,47 @@
+  /**
+   * Creates an anonymous sender.
+   *
+   * The sender will be established to the 'anonymous relay' and each message must specify its destination address.
```
This is largely duplicated by the next paragraphs first sentence, combine them?

gemmellr commented on src/main/java/io/vertx/ext/amqp/AmqpConnection.java:
```diff
@@ -79,19 +79,47 @@
+   * The sender will be established to the 'anonymous relay' and each message must specify its destination address.
+   *
+   * Unlike "regular" sender, this sender is not associated to a specific address, and each message sent must provide
+   * an address. This method is generally used in request-reply scenarios where you need a sender to send the reply,
```
I would remove 'This method is generally used...', as I dont think its entirely accurate or necessarily needed. Its common to use them for non request-response cases, or use fixed addresses for request-response.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpConnectionImpl.java:
```diff
@@ -265,20 +272,34 @@ public AmqpConnection receiver(String address, AmqpReceiverOptions receiverOptio
   @Override
-  public AmqpConnection sender(String address, Handler<AsyncResult<AmqpSender>> completionHandler) {
+  public AmqpConnection createSender(String address, Handler<AsyncResult<AmqpSender>> completionHandler) {
+    Objects.requireNonNull(address, "The address must be set");
```
I don't think I would require this. Is there a particular benefit?

gemmellr replied:
Ignore, you actually already changed it later to make it work with the other method, makes sense as it is now.

gemmellr commented on src/main/asciidoc/index.adoc:
```diff
@@ -141,6 +141,22 @@ as non-acknowledged (details can be found in the passed cause).
 
+== Implementing request-reply
+
+To implement a request-reply behavior, you need dynamic receiver and anonymous sender. A dynamic receiver is not associated with
```
This isnt necessarily true. Request-reply can also be done with fixed addresses and there are cases that is desirable and/or required.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpConnectionImpl.java:
```diff
@@ -333,17 +333,15 @@ public AmqpConnection createSender(String address, AmqpSenderOptions options, Ha
         sender = connection.get().createSender(address, opts);
         sender.setAutoDrained(options.isAutoDrained());
-        sender.setAutoDrained(options.isAutoSettle());
+        sender.setAutoSettle(options.isAutoSettle());
```
Exposing the ability to not locally 'auto settle' on send is a bit odd, since the API doesnt support manual settlement as alternative.

gemmellr commented on src/test/java/io/vertx/ext/amqp/SenderWithBrokerTest.java:
```diff
@@ -202,6 +204,76 @@ public void testSendBasicMessage(TestContext context) {
+        .setDynamic(true)
+        .setLinkName("my-link")
+        .setAutoDrained(false)
+        .setAutoSettle(false),
```
None of setLinkName("my-link").setAutoDrained(false).setAutoSettle(false) are really being used or tested here, so I would remove them.

gemmellr commented on src/test/java/io/vertx/ext/amqp/SenderWithBrokerTest.java:
```diff
@@ -202,6 +204,76 @@ public void testSendBasicMessage(TestContext context) {
+
+    await().until(() -> exec.get() != null);
+
+    ProtonClient proton = ProtonClient.create(vertx);
```
I think using ProtonServer rather than the broker+client would be a far better test here, so you can actually ensure a specific dynamic address allocated is returned from the address call. E.g. the equivalent functionality test from vertx-proton: https://github.com/vert-x3/vertx-proton/blob/3.6.3/src/test/java/io/vertx/proton/ProtonClientTest.java#L403

cescoffier committed:
Refactor receiver capabilities support

cescoffier committed:
Remove the ability to configure the TerminusDurability and TerminusExperiyPolicy

cescoffier committed:
Fix javadoc and documentation around request-reply

cescoffier committed:
Remove autosettle configuration

cescoffier committed:
Move test to use a mock server.

cescoffier committed:
Document how to create sender and receiver from the client directly

cescoffier commented:
@gemmellr I've addressed your comments.

gemmellr commented on src/test/java/io/vertx/ext/amqp/SenderTest.java:
```diff
@@ -334,4 +338,58 @@ private void doSenderClosedRemotelyCallsExceptionHandlerTestImpl(TestContext con
   }
+
+  @Test(timeout = 20000)
+  public void testDynamicSenderWithOptions(TestContext context) throws ExecutionException, InterruptedException {
```
Ah. Though my comment applies equally well to the sender case also, it was really the receiver case (since that's the commonly used variant rather than dynamic sender) which I meant to suggest would be better covered by a ProtonServer based test verifying handling of the dynamic address, like the one I then did actually link to. Oops :)

cescoffier replied:
I've added a test for the dynamic receiver case too.

gemmellr replied:
The new test doesn't validate that the address returned from the dynamic receiver is actually the one it sent from the server, it just checks non-null.

It also seems to takes nearly 11 seconds to complete, when I'd expect 0.x. From looking at the CI job, the overall test suite is rather slow which seems likely to be related (it seems slow even considering all the 3-4 sec periods burned on docker usage).

gemmellr replied:
I expect the delay here is because the test server doesn't have a connection close handler defined, so I guess the test is waiting on a close that never completes, with the 10 seconds seeming likely to be spent in BareTestBase tearDown.

In other tests in that class I think it might be waiting 10 seconds because the server was shut down after the client close started but before it completed.

cescoffier replied:
@gemmellr The delay is due to this:

```
receiver
              .closeHandler(done -> actualHandler.handle(done.mapEmpty()))
              .close();
```

The close handler is not called. Do you know what can be the reason?

gemmellr replied:
Which test? In any case, if it didnt fire then it presumably means it didn't actually receive a detach frame with closed=true. As I say, it looks like some of the tests perhaps stop their server a bit aggressively.

cescoffier replied:
I fixed most of them, 2 are remaining: 
* `io.vertx.ext.amqp.ReceiverCreditTest#testInitialCreditInfluencedByConsumerBufferSize`
* `io.vertx.ext.amqp.ReceiverCreditTest#testInitialCredit`

I believe that when the server is closed, the receiver cannot be closed afterward.

gemmellr replied:
To me it seems like those 2 tests close their test server in the test method, which will sever its connections, before then later closing the client in the teardown.

The latter looks to close the senders+receivers and then upon firing of their handlers proceed to close the underlying connection. It looks like if any of those never complete, neither does the overall process. None of them (inc connection) will ever close over the wire in this case since the connection was disconnected abruptly before the process even started.

From this Id say:
- The clients connection close handler should be accounting for the underlying connection disconnecting without close frame, both before it begins its close process and again after it has begun.
- Closing the senders and receivers in the connection close could probably just be dropped, a detach is implied when closing the connection, and doing so would simplify the whole process.

The other test in that class (which doesn't seem to have anything to do with credit hehe) doesn't shut down its server at all (though presumably the Vertx instance is closed at some point and does so) which will likely be be why it no longer delays after the other fix but these two still do.

cescoffier replied:
I'm going to remove the explicit closing of the receiver and sender qhen the connection is closed

cescoffier replied:
All fixed now!

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpConnectionImpl.java:
```diff
@@ -294,16 +294,6 @@ private void configureTheSource(AmqpReceiverOptions receiverOptions, ProtonRecei
     if (receiverOptions.isDurable()) {
       source.setExpiryPolicy(TerminusExpiryPolicy.NEVER);
       source.setDurable(TerminusDurability.UNSETTLED_STATE);
-    } else {
```
This only removed the use of the options, not ability to set the options which still remains.

cescoffier replied:
Fixed.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpConnectionImpl.java:
```diff
@@ -294,6 +289,8 @@ public AmqpConnection createReceiver(String address, AmqpReceiverOptions receive
     org.apache.qpid.proton.amqp.messaging.Source source = (org.apache.qpid.proton.amqp.messaging.Source) receiver
       .getSource();
+
+    source.setCapabilities(receiverOptions.getCapabilities().stream().map(Symbol::valueOf).toArray(Symbol[]::new));
```
This looks like it sets an empty array if there are no capabilities. It would be better not to call the setter in that case.

gemmellr replied:
Also, being able to change all this without any test failures probably indicates we need some tests.

cescoffier replied:
Because we update the tests accordingly. The CI didn't fail, but locally it has not been that peaceful.

gemmellr commented on src/main/java/io/vertx/ext/amqp/AmqpReceiverOptions.java:
```diff
@@ -115,50 +113,9 @@ public AmqpReceiverOptions setQos(String qos) {
 
   /**
-   * Gets the list of offered capabilities.
+   * Gets the list of offered capabilities for the source.
```
These aren't offered capabilities, they are more of an expectation for capabilities in this case. The peer has definitive say on what the capabilities of the Source are in the case of a client consuming (and similarly for the Target in case of a client sending).

Same applies to the 'setCapabilities' method below, not 'offered'.

gemmellr commented on src/main/java/io/vertx/ext/amqp/AmqpClientOptions.java:
```diff
@@ -0,0 +1,559 @@
+   * @param maxBufferSize the max buffered size, must be positive. If not set, default credit is used.
+   * @return the current {@link AmqpClientOptions} instance.
+   */
+  public AmqpClientOptions setMaxBufferedMessages(int maxBufferSize) {
```
This feels more like a receiver option.

cescoffier replied:
Agreed. It's a left-over from when I didn't have receiver options.

cescoffier committed:
Add a test to check dynamic receiver

cescoffier committed:
Do not set capabilities if none are set.

cescoffier committed:
Fix javadoc for the capabilities (offered replaced by desired)

cescoffier committed:
Move the maxBufferedMessages to the AMQP Receiver Options.

cescoffier committed:
Remove useless field
Make sure the copy constructor copy the maxBufferedMessages

cescoffier committed:
Remove options not used in code.

cescoffier commented:
@gemmellr I've done another pass.

vietj commented:
are we fine @gemmellr ?

gemmellr commented:
I added some comments last week after the most recent changes, there are some issues in the tests causing significant delays (which themselves may imply some issues in the actual code).

cescoffier commented:
@gemmellr Oh! I missed these comments.

vietj commented:
thanks for the heads-up !

> On 11 Mar 2019, at 17:00, Robbie Gemmell <notifications@github.com> wrote:
> 
> I added some comments last week after the most recent changes, there are some issues in the tests causing significant delays (which themselves may imply some issues in the actual code).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/vert-x3/vertx-amqp-client/pull/9#issuecomment-471603272>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AANxitf0cpj-5JfiIjc7I5jNNzylftzuks5vVn2pgaJpZM4bJMGA>.
>

cescoffier committed:
Reduce test duration by implementing close handler in the mock server

gemmellr commented:
Added another comment to above test etc discussion.

cescoffier committed:
Fix delay in tests
Set closed flag on disconnection
As discussed, do not close receivers and senders on connection close.

cescoffier commented:
@vietj @gemmellr I've fixed the tests.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpConnectionImpl.java:
```diff
@@ -0,0 +1,369 @@
+  }
+
+  @Override
+  public AmqpConnection close(Handler<AsyncResult<Void>> done) {
```
The close(..) process still isnt going to cater for a disconnection-without-close-response occurring after it starts.

In that case the handler passed to close(..) wont fire since no close frame arrives. Neither will the closeHandler(..) be fired for the same reason, but also that the underlying connections handler which did that work was replaced during close(..) and no longer does. The endHandler also seems like it won't be called in this scenario since 'closed' is set true during close(..) and the endHandler firing is gated on that. The last bit actually makes sense, but expectation is another handler is being called instead and it wont be here. Close(..) needs to make use of the disconnect handler. At least one of these 3 handlers should be called in that scenario.

Going back to an earlier comment from @vietj  about the closeHandler and endHandler both existing, I do now think change is needed there. Having a closeHandler(<handler>), endHandler(<handler>), and close(<handler>) is rather duplicative, I hadn't been taking note of close(..) having its own handler already, so if its going to then I also think one of the others should go away and their functionality be combined (i.e a handler that is called if either a close frame arrives or the connection disconnects without one, and close has not been called [with non-null handler?].

cescoffier replied:
I don't understand how to handle this case, would this be enough:
```
this.connection.get().closeHandler(pc -> {
      if (remoteCloseHandler != null) {
        runWithTrampoline(x -> {
          try {
            onDisconnect();
          } finally {
            onEnd();
            closed.set(true);
          }
          remoteCloseHandler.handle(this);
        });
      }
    });
```

Or should I also add a `disconnectHandler`:

```
Handler<AsyncResult<Void>> handler = pc -> {
      if (remoteCloseHandler != null) {
        runWithTrampoline(x -> {
          try {
            onDisconnect();
          } finally {
            onEnd();
            closed.set(true);
          }
          remoteCloseHandler.handle(this);
        });
      }
    };
    this.connection.get()
      .closeHandler(x -> handler.handle(x.mapEmpty()))
      .disconnectHandler(x -> handler.handle(Future.succeededFuture()));
```

You mention 3 handlers, but I only have disconnect and close. What's the third signals?

About the last part, that would mean removing the `endHandler`.

gemmellr replied:
The three handlers noted are from AmqpConnection:
AmqpConnection#close(\<handler\>);
AmqpConnection#closeHandler(\<handler\>);
AmqpConnection#endHandler(\<handler\>);

If the intent is to fire the AmqpConnection#close(\<handler\>) once that method is called, I think one of the other handlers goes away.

I think you should add a disconnectHandler as well as closeHandler to the underlying ProtonConnection within the AmqpConnection#close(\<handler\>) method impl. That way regardless whether the process completes normally, or if the underlying connection disconnects without a close frame being received, you get notified and can complete the handler passed to AmqpConnection#close(\<handler\>). The AmqpConnection#closeHandler(\<handler\>) and AmqpConnection#endHandler(\<handler\>) handlers can be documented not to fire after the point AmqpConnection#close(\<handler\>) returns. At which point removing one of them seems to make sense, and youve already somewhat started that by making AmqpConnection#closeHandler(\<handler\>) fire for disconnects as well as close frames that occur before AmqpConnection#close(\<handler\>).

Note that the underlying ProtonConnection closeHandler and disconnectHandler can/do both fire though, so you should guard completions etc on that. The former is purely indicating receipt of a close frame (either a response to locally-initiated close, or a remotely-initiated close), the latter the TCP disconnection, even if we asked for the closure. So its typically close+disconnect if a close frame is received, or just disconnect if a close frame is not received. I think a AmqpConnection#closeHandler(\<handler\>) being registered after the latest change would currently barf a little upon a server-initiated close.

cescoffier replied:
I've added an `exceptionHandler` replacing the `onEnd`(more inlined with the rest of Vert.x).
This handler is called on disconnect and unexpected closing (the server closes the connection).

In `close`, I register both `onDisconnect` and `onClose` handling the disconnect and close situation.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpConnectionImpl.java:
```diff
@@ -0,0 +1,369 @@
+  public AmqpConnection closeHandler(Handler<AmqpConnection> remoteCloseHandler) {
+    this.connection.get().closeHandler(pc -> {
+      if (remoteCloseHandler != null) {
+        runWithTrampoline(x -> remoteCloseHandler.handle(this));
```
As this replaces the underlying connections close handler directly, it should also run the onDisconnect bits like the default handler set while this object is created does. Thats need so that if this were a remotely-initiated close, rather than response to a locally-initiated close, then our side actually sends its close frame back (the protocol is symmetric).

cescoffier replied:
I've implemented the detachHandler.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpReceiverImpl.java:
```diff
@@ -0,0 +1,342 @@
+      handler(this.handler);
+    }
+
+    this.receiver.closeHandler(res -> {
```
This handles receipt of frame detach(closed=true), there is a related detachHandler for handling detach(closed=false)that should also be doing much the same thing here.

Need to vary whether to call receiver.close() or receiver.detach() as the AmqpReceiverImpl#close(..) method is doing.

cescoffier replied:
I've implemented the detachHandler.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpSenderImpl.java:
```diff
@@ -0,0 +1,258 @@
+    this.sender = sender;
+    this.connection = connection;
+
+    sender.closeHandler(res -> {
```
Similarly with the detachHandler().

cescoffier replied:
I've implemented the detachHandler.

gemmellr commented:
Expanded prior comment around the close(..) handling of disconnect a bit, still looks to be an issue there, and added some new comments on other niggles noticed while looking around again.

The precise trouble with the test delays seems fixed/sidestepped, and ones I could see were slow are running quickly now. From the CI output the overall job looks (I cant get the docker-using tests to work on either of my machines, and have given up trying) a good bit quicker now as a result which is nice, though still quite slow for all the tests there are.

cescoffier committed:
Handle detach events

cescoffier commented:
The slowness is mainly due to the docker image. The broker is started and stopped quite often and this takes time.

gemmellr commented:
Yep. Now seeing it in action (in CI at least) I'd probably remove the docker usage eventually, I don't think its adding much here except overhead. An embedded broker would be far faster. Most of the tests would be better tests of the client itself if they didn't use a broker.

cescoffier committed:
Add an exception handler on the connection to handle disconnect and unexpected (remote) closing.

cescoffier committed:
Connection can be null if the disconnect happen before the connection is established.

cescoffier commented:
I have been requested to test against other brokers, so the docker approach is pretty convenient for now.

gemmellr commented on src/main/java/io/vertx/ext/amqp/impl/AmqpConnectionImpl.java:
```diff
@@ -0,0 +1,373 @@
+                }
+              })
+              .closeHandler(x -> {
+                // Not expected closing, consider it failed
```
The client should actually close the connection on its end when the remote close arrives, to send the appropriate 'response' close frame back to the server, before then disconnecting the socket. It used to do this but it was removed in the latest changes.

The current calling of the new exception handler seems like it can occur multiple times for the same overall issue, e.g. once on unexpected remote close frame, then again on tcp disconnect. The prior endHandler stuff handled that by nulling out the callback when first firing it.

vietj commented:
@gemmellr @cescoffier can this PR be merged soon for vert.x 3.7 as tech preview ?

gemmellr commented:
Fine by me

cescoffier commented:
Woohoo! Merging!

cescoffier commented:
Thanks @gemmellr for the review!

vietj commented:
thanks @gemmellr it's a great step forward!

gemmellr commented:
Though I did expect my earlier feedback on issues with the latest changes (https://github.com/vert-x3/vertx-amqp-client/pull/9#discussion_r267148165) to be looked at before it was "merged soon".
